linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Instructions.scala

```diff
@@ -0,0 +1,395 @@
+
+  /** A stack-polymorphic instruction. */
+  sealed trait StackPolymorphicInstr extends Instr
+
+  /** An instruction that opens a structured control block. */
+  sealed trait StructuredLabeledInstr extends Instr {
+    val label: Option[LabelID]
+  }
+
+  // Convenience subclasses of instructions for writing text/binary
+
+  /** An instruction without any immediate argument. */
+  sealed abstract class SimpleInstr(mnemonic: String, opcode: Int)
+      extends Instr(mnemonic, opcode)
+
+  /** A structured labeled instruction with a single `BlockType` argument. */
+  sealed abstract class BlockTypeLabeledInstr(mnemonic: String, opcode: Int,
+      val blockTypeArgument: BlockType)
+      extends Instr(mnemonic, opcode) with StructuredLabeledInstr
+
+  /** An instruction with a single `LabelIdx` argument. */
+  sealed abstract class LabelInstr(mnemonic: String, opcode: Int,
+      val labelArgument: LabelID)
+      extends Instr(mnemonic, opcode)
+
+  /** An instruction with a single `FunctionID` argument. */
+  sealed abstract class FuncInstr(mnemonic: String, opcode: Int,
+      val funcArgument: FunctionID)
+      extends Instr(mnemonic, opcode)
+
+  /** An instruction with a single `TypeID` argument. */
+  sealed abstract class TypeInstr(mnemonic: String, opcode: Int,
+      val typeArgument: TypeID)
+      extends Instr(mnemonic, opcode)
+
+  /** An instruction with a single `TagID` argument. */
+  sealed abstract class TagInstr(mnemonic: String, opcode: Int,
+      val tagArgument: TagID)
+      extends Instr(mnemonic, opcode)
+
+  /** An instruction with a single `LocalID` argument. */
+  sealed abstract class LocalInstr(mnemonic: String, opcode: Int,
+      val localArgument: LocalID)
+      extends Instr(mnemonic, opcode)
+
+  /** An instruction with a single `GlobalID` argument. */
+  sealed abstract class GlobalInstr(mnemonic: String, opcode: Int,
+      val globalArgument: GlobalID)
+      extends Instr(mnemonic, opcode)
+
+  /** An instruction with a single `HeapType` argument. */
+  sealed abstract class HeapTypeInstr(mnemonic: String, opcode: Int,
+      val heapTypeArgument: HeapType)
+      extends Instr(mnemonic, opcode)
+
+  /** An instruction with a single `RefType` argument
+   *
+   *  In the binary format, it has split opcodes for the nullable and non-nullable variants.
+   */
+  sealed abstract class RefTypeInstr(mnemonic: String, nonNullOpcode: Int,
+      nullOpcode: Int, val refTypeArgument: RefType)
+      extends Instr(mnemonic, if (refTypeArgument.nullable) nullOpcode else nonNullOpcode)
+
+  /** An instruction with a pair of `TypeID`, `FieldID` arguments. */
+  sealed abstract class StructFieldInstr(mnemonic: String, opcode: Int,
+      val structTypeID: TypeID, val fieldID: FieldID)
+      extends Instr(mnemonic, opcode)
+
+  // The actual instruction list
+
+  // Fake instruction to mark position changes
+  final case class PositionMark(pos: Position) extends Instr("pos", -1)
+
+  // Unary operations
```

Comment:
Would it make sense to order all instruction definitions by opcode so they are easily referenceable in https://webassembly.github.io/gc/core/appendix/index-instructions.html ?

Reply:
Ah yes, that's probably a good idea. It grew somewhat organically, but the opcode order should be reasonably clean.